numba/types/abstract.py

```diff
@@ -209,9 +249,213 @@ class Hashable(Type):
     """
     Base class for hashable types.
     """
+    @classmethod
+    def interface_check(cls, instance):
```

Comment:
There is a problem, though: see how `Lt.interface_check()` has to know about every existing virtual subclass (here, only `UserLt`). This doesn't seem very extensible.

How about the following scheme (pseudo-code):
```
class Hashable:
    def is_hashable(self):
         # Default implementation
         return True
    def __instancecheck__(cls, inst):
         return issubclass(type(inst), cls) and inst.is_hashable()

class UserHashable(Hashable):
    def is_hashable(self):
         return self.supports_operator('hash')
```

Also, I think `test_types.py` should have the tests for all this.